File: commons/javafx/src/main/java/com/cedarsoft/commons/javafx/BindingsHelper.java
package com.cedarsoft.commons.javafx;

import java.text.NumberFormat;
import java.text.ParseException;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import javafx.beans.InvalidationListener;
import javafx.beans.binding.Binding;
import javafx.beans.binding.Bindings;
import javafx.beans.binding.LongBinding;
import javafx.beans.binding.StringBinding;
import javafx.beans.property.DoubleProperty;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.Property;
import javafx.beans.property.ReadOnlyIntegerProperty;
import javafx.beans.property.ReadOnlyProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableBooleanValue;
import javafx.beans.value.ObservableValue;
import javafx.beans.value.WritableDoubleValue;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.css.PseudoClass;
import javafx.scene.Node;
import javafx.scene.control.ComboBox;

public class BindingsHelper {
  private BindingsHelper() {
  }

  /**
   * Binds a combo box
   */
  @Nonnull
  public static <T extends Enum<T>> ComboBox<T> bindEnumCombo(@Nonnull ObjectProperty<T> property, @Nonnull Class<T> type) {
    return bindEnumCombo(new ComboBox<>(), property, type);
  }

  @Nonnull
  public static <T extends Enum<T>> ComboBox<T> bindEnumComboSpecificEnums(@Nonnull ObjectProperty<T> property, @Nonnull Class<T> type, @Nonnull T[] enumConstants) {
    return bindEnumComboSpecificEnums(new ComboBox<>(), property, type, enumConstants);
  }

  @Nonnull
  public static <T extends Enum<T>> ComboBox<T> bindEnumCombo(@Nonnull ComboBox<T> combo, @Nonnull ObjectProperty<T> property, @Nonnull Class<T> type) {
    return bindEnumComboSpecificEnums(combo, property, type, type.getEnumConstants());
  }

  @Nonnull
  public static <T extends Enum<T>> ComboBox<T> bindEnumComboSpecificEnums(@Nonnull ComboBox<T> combo, @Nonnull ObjectProperty<T> property, @Nonnull Class<T> type, @Nonnull T[] enumConstants) {
    combo.itemsProperty().set(FXCollections.observableArrayList(enumConstants));
    combo.valueProperty().bindBidirectional(property);

    combo.setButtonCell(new EnumListCell<>());
    combo.setCellFactory(param -> new EnumListCell<>());

    return combo;
  }



  @Nonnull
  public static StringBinding toBinding(@Nonnull StringProperty stringProperty) {
    return Bindings.createStringBinding(stringProperty::get, stringProperty);
  }

  public static <T extends Enum<T>> ComboBox<T> bindEnumCombo(@Nonnull ComboBox<T> combo, @Nonnull ObjectProperty<T> property, @Nonnull ObservableList<T> items) {
    combo.itemsProperty().set(items);
    combo.valueProperty().bindBidirectional(property);

    return combo;
  }

  /**
   * Sets the pseudo class when the given property is true
   */
  public static void bindPseudoClass(@Nonnull Node node, @Nonnull PseudoClass pseudoClass, @Nonnull ObservableBooleanValue pseudoClassActive) {
    //set initially
    node.pseudoClassStateChanged(pseudoClass, pseudoClassActive.get());

    //Update the pseudo class whenever the property changes
    pseudoClassActive.addListener((observable, oldValue, newValue) -> node.pseudoClassStateChanged(pseudoClass, newValue));
  }

  @Nonnull
  public static StringBinding asFormattedHex(@Nonnull ReadOnlyProperty<Number> longProperty) {
    return Bindings.createStringBinding(() -> {
      long longValue = longProperty.getValue().longValue();
      if (longValue == 0) {
        return "";
      }
      return toHex(longValue);
    }, longProperty);
  }

  /**
   * Converts to upper case hex with prefix "0x"
   */
  @Nonnull
  public static String toHex(long longValue) {
    return "0x" + Long.toHexString(longValue).toUpperCase();
  }

  /**
   * Wraps a string binding and ensures that no empty string is returned
   */
  @Nonnull
  public static Binding<String> avoidEmptyString(@Nonnull Binding<String> binding, @Nonnull String emptyStringReplacement) {
    return new Binding<String>() {
      @Override
      public String getValue() {
        String value = binding.getValue();
        if (value == null || value.trim().isEmpty()) {
          return emptyStringReplacement;
        }

        return value;
      }

      @Override
      public boolean isValid() {
        return binding.isValid();
      }

      @Override
      public void invalidate() {
        binding.invalidate();
      }

      @Override
      public ObservableList<?> getDependencies() {
        return binding.getDependencies();
      }

      @Override
      public void dispose() {
        binding.dispose();
      }

      @Override
      public void addListener(ChangeListener<? super String> listener) {
        binding.addListener(listener);
      }

      @Override
      public void removeListener(ChangeListener<? super String> listener) {
        binding.removeListener(listener);
      }

      @Override
      public void addListener(InvalidationListener listener) {
        binding.addListener(listener);
      }

      @Override
      public void removeListener(InvalidationListener listener) {
        binding.removeListener(listener);
      }
    };
  }

  @Nonnull
  public static StringBinding bindFormattedNumber(@Nonnull ObservableValue<Number> property, @Nonnull NumberFormat numberFormat) {
    return Bindings.createStringBinding(() -> numberFormat.format(property.getValue()), property);
  }

  /**
   * Creates a listener that ensures a given property always is within the given min and max value (both inclusive)
   */
  @Nonnull
  public static ChangeListener<Number> createBoundsChecker(double minValue, double maxValue) {
    return (observable, oldValue, newValue) -> {
      if (newValue.doubleValue() < minValue) {
        ((WritableDoubleValue) observable).set(minValue);
      }
      if (newValue.doubleValue() > maxValue) {
        ((WritableDoubleValue) observable).set(maxValue);
      }
    };
  }

  /**
   * Can be used to bind a double property to a string property
   */
  @Nonnull
  public static Property<String> double2stringBiDirectional(@Nonnull DoubleProperty doubleProperty) {
    SimpleStringProperty stringProperty = new SimpleStringProperty();

    stringProperty.addListener((observable, oldValue, newValue) -> {
      try {
        double doubleValue = string2double(newValue);
        doubleProperty.setValue(doubleValue);
      } catch (ParseException e) {
        //Enforce the old values if can not be parsed
        stringProperty.setValue(oldValue);
      }
    });

    doubleProperty.addListener(new ChangeListener<Number>() {
      @Override
      public void changed(ObservableValue<? extends Number> observable, Number oldValue, Number newValue) {
        try {
          double oldValueFromString = string2double(stringProperty.get());
          if (oldValueFromString == newValue.doubleValue()) {
            return;
          }
        } catch (ParseException ignore) {
        }

        stringProperty.setValue(NumberFormat.getInstance().format(newValue.doubleValue()));
      }
    });

    return stringProperty;
  }

  /**
   * Converts a string to double. Handles null values and empty strings
   */
  private static double string2double(@Nullable String newValue) throws ParseException {
    if (newValue == null || newValue.isEmpty()) {
      return 0.0;
    }

    return NumberFormat.getInstance().parse(newValue.trim()).doubleValue();
  }

  /**
   * creates a LongBinding from three Integer properties.
   * @param intProperty0
   * @param intProperty1
   * @param intProperty2
   */
  public static LongBinding createLongBinding(@Nonnull ReadOnlyIntegerProperty intProperty0,
                                              @Nonnull ReadOnlyIntegerProperty intProperty1,
                                              @Nonnull ReadOnlyIntegerProperty intProperty2) {

    return Bindings.createLongBinding(() -> {

      Integer byte0 = intProperty0.getValue() & 0xffffff;
      Integer byte1 = intProperty1.getValue() & 0xffff;
      Integer byte2 = intProperty2.getValue() & 0xff;

      return (long) (byte2 + byte1 * 256 + byte0 * 256 * 256);
    }, intProperty0, intProperty1, intProperty2);
  }

  /**
   * Binds two method calls to a boolean property.
   * The method corresponding to the current state is called once initially
   */
  public static void bindMethodsToBoolean(@Nonnull ObservableValue<Boolean> property, @Nonnull Runnable onTrue, @Nonnull Runnable onFalse) {
    property.addListener((observable, oldValue, newValue) -> {
      if (newValue) {
        onTrue.run();
      }
      else {
        onFalse.run();
      }
    });

    //Initial
    if (property.getValue()) {
      onTrue.run();
    }
    else {
      onFalse.run();
    }
  }
}

File: commons/javafx/src/main/kotlin/com/cedarsoft/commons/javafx/BindingsHelper.kt
package com.cedarsoft.commons.javafx

import com.cedarsoft.commons.javafx.properties.*
import javafx.beans.InvalidationListener
import javafx.beans.binding.Binding
import javafx.beans.binding.Bindings
import javafx.beans.binding.LongBinding
import javafx.beans.binding.StringBinding
import javafx.beans.binding.StringExpression
import javafx.beans.property.DoubleProperty
import javafx.beans.property.ObjectProperty
import javafx.beans.property.Property
import javafx.beans.property.ReadOnlyIntegerProperty
import javafx.beans.property.ReadOnlyProperty
import javafx.beans.property.SimpleStringProperty
import javafx.beans.value.ChangeListener
import javafx.beans.value.ObservableBooleanValue
import javafx.beans.value.ObservableValue
import javafx.beans.value.WritableDoubleValue
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.css.PseudoClass
import javafx.scene.Node
import javafx.scene.control.ComboBox
import javafx.scene.control.ListView
import java.text.NumberFormat
import java.text.ParseException
import java.util.Locale

object BindingsHelper {
  /**
   * Binds a combo box
   */
  @JvmStatic
  fun <T : Enum<T>> bindEnumCombo(property: ObjectProperty<T>, type: Class<T>): ComboBox<T> {
    return bindEnumCombo(ComboBox(), property, type)
  }

  @JvmStatic
  fun <T : Enum<T>> bindEnumCombo(property: PropertyWithWritableState<T>, type: Class<T>): ComboBox<T> {
    return bindEnumCombo(ComboBox(), property, type)
  }

  @JvmStatic
  fun <T : Enum<T>> bindEnumComboSpecificEnums(property: Property<T>, type: Class<T>, enumConstants: Array<T>): ComboBox<T> {
    return bindEnumComboSpecificEnums(ComboBox(), property, type, enumConstants)
  }

  @JvmStatic
  fun <T : Enum<T>> bindEnumComboSpecificEnums(property: PropertyWithWritableState<T>, type: Class<T>, enumConstants: Array<T>): ComboBox<T> {
    return bindEnumComboSpecificEnums(ComboBox(), property, type, enumConstants)
  }

  @JvmStatic
  fun <T : Enum<T>> bindEnumCombo(combo: ComboBox<T>, property: ObjectProperty<T>, type: Class<T>): ComboBox<T> {
    return bindEnumComboSpecificEnums(combo, property, type, type.enumConstants)
  }

  @JvmStatic
  fun <T : Enum<T>> bindEnumCombo(combo: ComboBox<T>, property: PropertyWithWritableState<T>, type: Class<T>): ComboBox<T> {
    return bindEnumComboSpecificEnums(combo, property, type, type.enumConstants)
  }

  @JvmStatic
  fun <T : Enum<T>> bindEnumComboSpecificEnums(combo: ComboBox<T>, property: PropertyWithWritableState<T>, type: Class<T>, enumConstants: Array<T>): ComboBox<T> {
    bindEnumComboSpecificEnums(combo, property.property, type, enumConstants)
    combo.disableProperty().bind(property.writableProperty.not())
    return combo
  }

  @JvmStatic
  fun <T : Enum<T>> bindEnumComboSpecificEnums(combo: ComboBox<T>, property: Property<T>, type: Class<T>, enumConstants: Array<T>): ComboBox<T> {
    combo.itemsProperty().set(FXCollections.observableArrayList(*enumConstants))
    combo.valueProperty().bindBidirectional(property)
    combo.setButtonCell(EnumListCell())
    combo.setCellFactory { param: ListView<T> -> EnumListCell() }
    return combo
  }

  @JvmStatic
  fun toBinding(stringProperty: StringExpression): StringBinding {
    return Bindings.createStringBinding({ stringProperty.get() }, stringProperty)
  }

  @JvmStatic
  fun <T : Enum<T>> bindEnumCombo(combo: ComboBox<T>, property: ObjectProperty<T>, items: ObservableList<T>): ComboBox<T> {
    combo.itemsProperty().set(items)
    combo.valueProperty().bindBidirectional(property)
    return combo
  }

  /**
   * Sets the pseudo class when the given property is true
   */
  @JvmStatic
  fun bindPseudoClass(node: Node, pseudoClass: PseudoClass, pseudoClassActive: ObservableBooleanValue) {
    //set initially
    node.pseudoClassStateChanged(pseudoClass, pseudoClassActive.get())

    //Update the pseudo class whenever the property changes
    pseudoClassActive.addListener { observable: ObservableValue<out Boolean?>, oldValue: Boolean?, newValue: Boolean? -> node.pseudoClassStateChanged(pseudoClass, newValue!!) }
  }

  @JvmStatic
  fun asFormattedHex(longProperty: ReadOnlyProperty<Number>): StringBinding {
    return Bindings.createStringBinding({
      val longValue = longProperty.value.toLong()
      if (longValue == 0L) {
        return@createStringBinding ""
      }
      toHex(longValue)
    }, longProperty)
  }

  /**
   * Converts to upper case hex with prefix "0x"
   */
  @JvmStatic
  fun toHex(longValue: Long): String {
    return "0x" + java.lang.Long.toHexString(longValue).uppercase(Locale.getDefault())
  }

  /**
   * Wraps a string binding and ensures that no empty string is returned
   */
  @JvmStatic
  fun avoidEmptyString(binding: Binding<String>, emptyStringReplacement: String): Binding<String> {
    return object : Binding<String> {
      override fun getValue(): String {
        val value = binding.value
        return if (value == null || value.trim { it <= ' ' }.isEmpty()) {
          emptyStringReplacement
        } else value
      }

      override fun isValid(): Boolean {
        return binding.isValid
      }

      override fun invalidate() {
        binding.invalidate()
      }

      override fun getDependencies(): ObservableList<*> {
        return binding.dependencies
      }

      override fun dispose() {
        binding.dispose()
      }

      override fun addListener(listener: ChangeListener<in String>) {
        binding.addListener(listener)
      }

      override fun removeListener(listener: ChangeListener<in String>) {
        binding.removeListener(listener)
      }

      override fun addListener(listener: InvalidationListener) {
        binding.addListener(listener)
      }

      override fun removeListener(listener: InvalidationListener) {
        binding.removeListener(listener)
      }
    }
  }

  @JvmStatic
  fun bindFormattedNumber(property: ObservableValue<Number>, numberFormat: NumberFormat): StringBinding {
    return Bindings.createStringBinding({
      val value = property.value
      if (java.lang.Double.isNaN(value.toDouble())) {
        return@createStringBinding "-"
      }
      numberFormat.format(value)
    }, property)
  }

  /**
   * Creates a listener that ensures a given property always is within the given min and max value (both inclusive)
   */
  @JvmStatic
  fun createBoundsChecker(minValue: Double, maxValue: Double): ChangeListener<Number> {
    return ChangeListener { observable: ObservableValue<out Number>, oldValue: Number?, newValue: Number ->
      if (newValue.toDouble() < minValue) {
        (observable as WritableDoubleValue).set(minValue)
      }
      if (newValue.toDouble() > maxValue) {
        (observable as WritableDoubleValue).set(maxValue)
      }
    }
  }

  /**
   * Can be used to bind a double property to a string property
   */
  @JvmStatic
  fun double2stringBiDirectional(doubleProperty: DoubleProperty): Property<String> {
    val stringProperty = SimpleStringProperty()
    stringProperty.addListener { observable: ObservableValue<out String?>, oldValue: String?, newValue: String? ->
      try {
        val doubleValue = string2double(newValue)
        doubleProperty.value = doubleValue
      } catch (e: ParseException) {
        //Enforce the old values if can not be parsed
        stringProperty.value = oldValue
      }
    }
    doubleProperty.addListener(ChangeListener { observable, oldValue, newValue ->
      try {
        val oldValueFromString = string2double(stringProperty.get())
        if (oldValueFromString == newValue.toDouble()) {
          return@ChangeListener
        }
      } catch (ignore: ParseException) {
      }
      stringProperty.value = NumberFormat.getInstance().format(newValue.toDouble())
    })
    return stringProperty
  }

  /**
   * Converts a string to double. Handles null values and empty strings
   */
  @Throws(ParseException::class)
  private fun string2double(newValue: String?): Double {
    return if (newValue.isNullOrEmpty()) {
      0.0
    } else NumberFormat.getInstance().parse(newValue.trim { it <= ' ' }).toDouble()
  }

  /**
   * creates a LongBinding from three Integer properties.
   * @param intProperty0
   * @param intProperty1
   * @param intProperty2
   */
  @JvmStatic
  fun createLongBinding(
    intProperty0: ReadOnlyIntegerProperty,
    intProperty1: ReadOnlyIntegerProperty,
    intProperty2: ReadOnlyIntegerProperty,
  ): LongBinding {
    return Bindings.createLongBinding({
      val byte0 = intProperty0.value and 0xffffff
      val byte1 = intProperty1.value and 0xffff
      val byte2 = intProperty2.value and 0xff
      (byte2 + byte1 * 256 + byte0 * 256 * 256).toLong()
    }, intProperty0, intProperty1, intProperty2)
  }

  /**
   * Binds two method calls to a boolean property.
   * The method corresponding to the current state is called once initially
   */
  @JvmStatic
  fun bindMethodsToBoolean(property: ObservableValue<Boolean>, onTrue: Runnable, onFalse: Runnable) {
    property.addListener { _: ObservableValue<out Boolean>?, _: Boolean?, newValue: Boolean ->
      if (newValue) {
        onTrue.run()
      } else {
        onFalse.run()
      }
    }

    //Initial
    if (property.value) {
      onTrue.run()
    } else {
      onFalse.run()
    }
  }
}
